LOOP-1328: Periodically store eligible Glucose Samples into HealthKit …

rickpasetto opened the pull request:
[LOOP-1328](https://tidepool.atlassian.net/browse/LOOP-1328)
- instead of storing them "in line", does a query for "eligible" (i.e. not too recent) samples that have not been saved to HK yet (i.e. 'uuid = nil') and then writes them
- only does this, for now, on every new sample
- also: Adds `device` to `CachedGlucoseObject`

Commit:
LOOP-1328: Periodically store eligible Glucose Samples into HealthKit

- instead of storing them "in line", does a query for "eligible" (i.e. not too recent) samples that have not been saved to HK yet (i.e. 'uuid = nil') and then writes them
- only does this, for now, on every new sample, but probably better done on a regular basis(?)
- still an open question about the HKDevice, which is lost during this process, because we don't store it in CoreData.

Commit:
LOOP-1328: Adds `device` to CachedGlucoseObject

Commit:
Restore uuid to StoredGlucoseSample

ps2 commented on LoopKit/Persistence/Model.xcdatamodeld/Model.xcdatamodel/contents:
```diff
@@ -1,16 +1,16 @@
-        <attribute name="syncVersion" attributeType="Integer 32" defaultValueString="1" usesScalarValueType="YES" syncable="YES"/>
-        <attribute name="uploadState" attributeType="Integer 32" defaultValueString="0" usesScalarValueType="NO" syncable="YES"/>
-        <attribute name="uuid" optional="YES" attributeType="UUID" usesScalarValueType="NO" syncable="YES"/>
+        <attribute name="absorptionTime" optional="YES" attributeType="Double" usesScalarValueType="NO"/>
```
There is a pretty significant data migration happening in the DIY sync. Can this go in after that lands?

rickpasetto replied:
Yep yep!  I'll hold off.  (I also still have some extant changes I'm working on in `Loop` anyway...)

ps2 commented on MockKit/MockCGMManager.swift:
```diff
@@ -450,6 +450,8 @@ public final class MockCGMManager: TestingCGMManager {
 
     public let shouldSyncToRemoteService = false
     
+    public static let healthKitStorageDelay: TimeInterval = .minutes(5)
```
Would be nice to have this configurable. Not a blocker.

rickpasetto replied:
Since it is `static` you would have to delete and re-add the CGM simulator, but at least that's doable.

ps2 reviewed (approved):
LGTM. I'd like to get the DIY migration in place first, if that's ok, since it's been tested, and is close to landing.

Commit:
HKDevice+Encodable

Commit:
PR Feedback: configurable healthKitStorageDelay
(NOTE: you have to delete and re-add the CGM simulator for it to take effect though!)

rickpasetto commented:
@ps2 note the addition of `HKDevice+Encodable` in order to support "Codable" in HKDevice.

Commit:
Merge branch 'dev' into rick/LOOP-1328

Commit:
Moved added `device` attribute to v4

rickpasetto commented:
@ps2 ok, updated post DIY-sync.  I moved the added `device` attribute to the v4 Model... I'm not sure, do I also have to update the mapping model?

Commit:
Fix merge error, add HKDeviceCodable unit tests

Commit:
Update unit tests

ps2 commented:
Yes, the mapping model needs to be updated, if it hasn't been already. You can't just recreate it. After the model is updated, restart xcode, and then select the Model1v4 mapping, and Editor -> Refresh Models.  I think that will update the current mapping. Make sure it looks good for the new attribute, and it should be fine.

ps2 commented on LoopKit/GlucoseKit/CachedGlucoseObject+CoreDataClass.swift:
```diff
@@ -21,7 +21,20 @@ class CachedGlucoseObject: NSManagedObject {
+        get {
+            willAccessValue(forKey: "device")
+            defer { didAccessValue(forKey: "device") }
+            return primitiveDevice.flatMap { try? NSKeyedUnarchiver.unarchivedObject(ofClass: HKDevice.self, from: $0) }
```
Serializing the device into a relational column isn't great. We did it in DosingDecisionStore, as there Loop specific things that don't belong in LoopKit, and we don't imagine querying dosing decisions in the near term. I think you're right that this would be a good candidate for denormalization; especially because we'll likely just keep storing the same device over and over. However, there isn't a strict need for us to query this column yet, so I won't block this because of that.

rickpasetto replied:
I was a bit concerned about this as well.  And I came up with the same conclusion: this is just stuffed in CoreData only for rehydration for HealthKit (well, also for uploading to Services if we want).  But we kinda have no choice, we do have to store it (IMHO it is a good thing that the device details are available to the user in Apple Health). Point is: we don't query or index on it, so I thought a serialized format was fine.  I measured, for example, the size of Dexcom's `HKDevice`, and it winds up being ~400 bytes.  Not gigantic, but not tiny either.  If we want to optimize this, we should first figure out what we want to optimize (database size? processing speed? upload bandwidth? something else?).  If we decide size needs to be optimized, I think denormalization would be pretty straightforward.

Commit:
Update Modelv1v4 Entity Mappings

Commit:
add CoreDataMigrationTests

Commit:
Merge dev in
